File: DynamicKey/AgoraDynamicKey/dart/test/stt_token_builder_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:agora_token_generator/agora_token_generator.dart';

void main() {
  const appId = '970CA35de60c44645bbae8a215061b33';
  const appCertificate = '5CFd2fd1755d40ecb72977518be15d3b';
  const channelName = 'stt-channel';
  const rtcAccount = 'stt-rtc-user';
  const rtmUserId = 'stt-rtm-user';
  const rtcTokenExpireSeconds = 3600;
  const joinChannelPrivilegeExpire = 1800;
  const publishAudioPrivilegeExpire = 1700;
  const publishVideoPrivilegeExpire = 1600;
  const publishDataPrivilegeExpire = 1500;
  const rtmTokenExpireSeconds = 1400;

  test('builds a publisher STT token', () {
    final token = SttTokenBuilder.buildToken(
      appId: appId,
      appCertificate: appCertificate,
      channelName: channelName,
      rtcAccount: rtcAccount,
      rtcRole: RtcRole.publisher,
      rtcTokenExpireSeconds: rtcTokenExpireSeconds,
      joinChannelPrivilegeExpire: joinChannelPrivilegeExpire,
      publishAudioPrivilegeExpire: publishAudioPrivilegeExpire,
      publishVideoPrivilegeExpire: publishVideoPrivilegeExpire,
      publishDataPrivilegeExpire: publishDataPrivilegeExpire,
      rtmUserId: rtmUserId,
      rtmTokenExpireSeconds: rtmTokenExpireSeconds,
    );

    final parsed = AccessToken.empty();
    expect(parsed.parse(token), isTrue);
    final rtc = parsed.getServices(Service.RTC).single as ServiceRtc;
    final rtm = parsed.getServices(Service.RTM).single as ServiceRtm;
    final stt = parsed.getServices(Service.STT).single as ServiceStt;
    expect(parsed.appId, appId);
    expect(parsed.expire, rtcTokenExpireSeconds);
    expect(rtc.channelName, channelName);
    expect(rtc.uid, rtcAccount);
    expect(rtc.privileges, {
      ServiceRtc.privilegeJoinChannel: joinChannelPrivilegeExpire,
      ServiceRtc.privilegePublishAudioStream: publishAudioPrivilegeExpire,
      ServiceRtc.privilegePublishVideoStream: publishVideoPrivilegeExpire,
      ServiceRtc.privilegePublishDataStream: publishDataPrivilegeExpire,
    });
    expect(rtm.userId, rtmUserId);
    expect(rtm.privileges[ServiceRtm.privilegeLogin], rtmTokenExpireSeconds);
    expect(stt.privileges, isEmpty);
  });

  test('builds a subscriber STT token', () {
    final token = SttTokenBuilder.buildToken(
      appId: appId,
      appCertificate: appCertificate,
      channelName: channelName,
      rtcAccount: rtcAccount,
      rtcRole: RtcRole.subscriber,
      rtcTokenExpireSeconds: rtcTokenExpireSeconds,
      joinChannelPrivilegeExpire: joinChannelPrivilegeExpire,
      publishAudioPrivilegeExpire: publishAudioPrivilegeExpire,
      publishVideoPrivilegeExpire: publishVideoPrivilegeExpire,
      publishDataPrivilegeExpire: publishDataPrivilegeExpire,
      rtmUserId: rtmUserId,
      rtmTokenExpireSeconds: rtmTokenExpireSeconds,
    );

    final parsed = AccessToken.empty();
    expect(parsed.parse(token), isTrue);
    final rtc = parsed.getServices(Service.RTC).single as ServiceRtc;
    expect(rtc.privileges, {
      ServiceRtc.privilegeJoinChannel: joinChannelPrivilegeExpire,
    });
  });
}

File: DynamicKey/AgoraDynamicKey/nodejs/test/SttTokenBuilderTest.js
/**
 * run this test with command:
 * nodeunit test/SttTokenBuilderTest.js
 */
const { SttTokenBuilder } = require('../src/SttTokenBuilder')
const { AccessToken2, kRtcServiceType, kRtmServiceType, kSttServiceType, ServiceRtc, ServiceRtm } = require('../src/AccessToken2')
const { Role } = require('../src/RtcTokenBuilder2')

const appId = '970CA35de60c44645bbae8a215061b33'
const appCertificate = '5CFd2fd1755d40ecb72977518be15d3b'
const channelName = '7d72365eb983485397e3e3f9d460bdda'
const rtcAccount = '2882341273'
const rtmUserId = '2882341273'
const rtcTokenExpire = 600
const joinChannelPrivilegeExpire = 600
const pubAudioPrivilegeExpire = 600
const pubVideoPrivilegeExpire = 600
const pubDataStreamPrivilegeExpire = 600
const rtmTokenExpire = 600

exports.buildToken_PUBLISHER_Test = function (test) {
    const token = SttTokenBuilder.buildToken(
        appId, appCertificate, channelName, rtcAccount, Role.PUBLISHER, rtcTokenExpire,
        joinChannelPrivilegeExpire, pubAudioPrivilegeExpire, pubVideoPrivilegeExpire,
        pubDataStreamPrivilegeExpire, rtmUserId, rtmTokenExpire)
    const parsed = new AccessToken2('', '', 0, 0)
    parsed.from_string(token)

    const rtc = parsed.getServices(kRtcServiceType)[0]
    const rtm = parsed.getServices(kRtmServiceType)[0]
    const stt = parsed.getServices(kSttServiceType)[0]

    test.equal(appId, parsed.appId)
    test.equal(rtcTokenExpire, parsed.expire)
    test.equal(channelName, rtc.__channel_name)
    test.equal(rtcAccount, rtc.__uid)
    test.equal(joinChannelPrivilegeExpire, rtc.__privileges[ServiceRtc.kPrivilegeJoinChannel])
    test.equal(pubAudioPrivilegeExpire, rtc.__privileges[ServiceRtc.kPrivilegePublishAudioStream])
    test.equal(pubVideoPrivilegeExpire, rtc.__privileges[ServiceRtc.kPrivilegePublishVideoStream])
    test.equal(pubDataStreamPrivilegeExpire, rtc.__privileges[ServiceRtc.kPrivilegePublishDataStream])
    test.equal(rtmUserId, rtm.__user_id)
    test.equal(rtmTokenExpire, rtm.__privileges[ServiceRtm.kPrivilegeLogin])
    test.deepEqual({}, stt.__privileges)
    test.done()
}

exports.buildToken_SUBSCRIBER_Test = function (test) {
    const token = SttTokenBuilder.buildToken(
        appId, appCertificate, channelName, rtcAccount, Role.SUBSCRIBER, rtcTokenExpire,
        joinChannelPrivilegeExpire, pubAudioPrivilegeExpire, pubVideoPrivilegeExpire,
        pubDataStreamPrivilegeExpire, rtmUserId, rtmTokenExpire)
    const parsed = new AccessToken2('', '', 0, 0)
    parsed.from_string(token)

    const rtc = parsed.getServices(kRtcServiceType)[0]
    test.deepEqual({ [ServiceRtc.kPrivilegeJoinChannel]: joinChannelPrivilegeExpire }, rtc.__privileges)
    test.done()
}
